httpcore/_async/http2.py

```diff
@@ -314,17 +314,41 @@ async def send_headers(
     ) -> None:
         scheme, hostname, port, path = url
         default_port = {b"http": 80, b"https": 443}.get(scheme)
-        if port is None or port == default_port:
+
+        # In HTTP/2 the ':authority' pseudo-header is used instead of 'Host'.
+        # For convenience we automatically treat any 'Host' header value
+        # as a value for ':authority', allowing users to pass 'Host' whenever they
+        # want to customize the host header regardless of the HTTP version in use.
+        authority = None
+        for k, v in headers:
+            if k == b"host":
+                authority = v
+                break
+
+        if authority is None:
             authority = hostname
-        else:
-            authority = b"%s:%d" % (hostname, port)
+
+        if port is not None and port != default_port:
+            authority = b"%s:%d" % (authority, port)
 
         headers = [
             (b":method", method),
             (b":authority", authority),
             (b":scheme", scheme),
             (b":path", path),
-        ] + [(k, v) for k, v in headers if k not in (b"host", b"transfer-encoding")]
+        ] + [
+            (k, v)
+            for k, v in headers
+            if k
+            not in (
+                b"host",
+                b"transfer-encoding",
+                b":method",
+                b":authority",
+                b":scheme",
+                b":path",
+            )
+        ]
```

Comment:
I'd prefer if we didn't change the pseudo-header behaviour as part of this PR.

There's some complexities here that'd need further discussion. Since this change isn't necessary for the rest of the pull request to make sense, let's treat anything there independantly.

Reply:
I agree that it isn't part of this fix, but I changed it because I tried adding ":authority" to the headers and got an exception because of duplicate headers.

Reply:
Sure thing. If we want to tackle that let's do so separately tho.

For instance, if we are going to strip pseudo headers like this, then we'd probably also want to do the same for HTTP/1.1 codepaths.
 
It's part of a more general conversation about "which header values should be treated as valid".